<a href="https://colab.research.google.com/github/mossytreesandferns/DataScience2/blob/master/MulticlassClassificationFourApproaches.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Mount my drive- run the code, go to the link, accept.
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/gdrive


In [ ]:
#Change working directory to make it easier to access the files
import os
os.chdir("/content/gdrive/My Drive/Colab Notebooks/")
os.getcwd()

'/content/gdrive/My Drive/Colab Notebooks'

# Multiclass Classification in Four Approaches

In [ ]:
# Import Libraries
import numpy as np
import seaborn as sns
import pandas as pd

# Graphs 
import matplotlib.pyplot as plt

# For these examples, sklearn is featured
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


### From the Forest Cover Type Data Set, accessed on Kaggle, we can try to predict the forest cover type in the Roosevelt National Forest in Colorado.  The data set contains fifty-four input variables including, elevation, slope, aspect, and horizontal and vertical distance to water.  There are seven possible output labels, or tree species cover.  These are: 
### 1. Spruce/Fir    2. Lodgepole Pine    3. Ponderosa Pine    4. Cottonwood/Willow    5. Aspen    6. Douglas-fir and 7. Krummholz.

##### Complete information can be found here: https://www.kaggle.com/uciml/forest-cover-type-dataset

### Preliminary Look at Data

In [ ]:
forest_raw = pd.read_csv('/content/gdrive/My Drive/OwnJupyterNotebooks/covtype.csv')
forest_raw.head()

Elevation  Aspect  Slope  ...  Soil_Type39  Soil_Type40  Cover_Type
0       2596      51      3  ...            0            0           5
1       2590      56      2  ...            0            0           5
2       2804     139      9  ...            0            0           2
3       2785     155     18  ...            0            0           2
4       2595      45      2  ...            0            0           5

[5 rows x 55 columns]

In [ ]:
# Check for null values
forest_raw.isnull().sum()

Elevation                             0
Aspect                                0
Slope                                 0
Horizontal_Distance_To_Hydrology      0
Vertical_Distance_To_Hydrology        0
Horizontal_Distance_To_Roadways       0
Hillshade_9am                         0
Hillshade_Noon                        0
Hillshade_3pm                         0
Horizontal_Distance_To_Fire_Points    0
Wilderness_Area1                      0
Wilderness_Area2                      0
Wilderness_Area3                      0
Wilderness_Area4                      0
Soil_Type1                            0
Soil_Type2                            0
Soil_Type3                            0
Soil_Type4                            0
Soil_Type5                            0
Soil_Type6                            0
Soil_Type7                            0
Soil_Type8                            0
Soil_Type9                            0
Soil_Type10                           0
Soil_Type11                           0


## Decision Tree Classifier

In [ ]:
# X is the classification predictor characteristics, y is the target label
X = forest_raw[['Elevation','Aspect','Slope','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology',
                'Horizontal_Distance_To_Roadways','Hillshade_9am','Hillshade_Noon','Hillshade_3pm','Horizontal_Distance_To_Fire_Points',
                'Wilderness_Area1','Wilderness_Area2','Wilderness_Area3','Wilderness_Area4','Soil_Type1','Soil_Type2',
                'Soil_Type3','Soil_Type4','Soil_Type5','Soil_Type6','Soil_Type7','Soil_Type8','Soil_Type9','Soil_Type10','Soil_Type11',
                'Soil_Type12','Soil_Type13','Soil_Type14','Soil_Type15','Soil_Type16','Soil_Type17','Soil_Type18','Soil_Type19',
                'Soil_Type20','Soil_Type21','Soil_Type22','Soil_Type23','Soil_Type24','Soil_Type25','Soil_Type26',
                'Soil_Type27','Soil_Type28','Soil_Type29','Soil_Type30','Soil_Type31','Soil_Type32','Soil_Type33',
                'Soil_Type34','Soil_Type35','Soil_Type36','Soil_Type37','Soil_Type38','Soil_Type39','Soil_Type40']]
y = forest_raw['Cover_Type']

In [ ]:
# Divide X and y into training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

In [ ]:
# Train a DescisionTreeClassifier 
from sklearn.tree import DecisionTreeClassifier 
decision_tree_model = DecisionTreeClassifier(max_depth = 2).fit(X_train, y_train) 
decision_tree_predictions = decision_tree_model.predict(X_test) 

In [ ]:
decision_tree_predictions

array([2, 2, 3, ..., 2, 2, 1])

### What is the accuracy of the Decision Tree Model on this Data Set?

In [ ]:
# Get a measure of accuracy
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test, decision_tree_predictions)
print(accuracy)

0.6703200622362361


#### The accuracy is 0.67

### A more complete report

In [ ]:
# SKLEARN classification report
from sklearn.metrics import classification_report
report = classification_report(y_test, decision_tree_predictions)
print(report)

              precision    recall  f1-score   support

           1       0.63      0.73      0.68     52943
           2       0.72      0.74      0.73     70897
           3       0.57      0.69      0.62      8890
           4       0.00      0.00      0.00       704
           5       0.00      0.00      0.00      2340
           6       0.00      0.00      0.00      4384
           7       0.00      0.00      0.00      5095

    accuracy                           0.67    145253
   macro avg       0.27      0.31      0.29    145253
weighted avg       0.62      0.67      0.64    145253



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## Support Vector Machines (SVM) Classification

In [ ]:
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split

In [ ]:
# Read Data
forest = pd.read_csv('/content/gdrive/My Drive/OwnJupyterNotebooks/covtype.csv')

In [ ]:
# X is the classification predictor characteristics, y is the target label
X = forest[['Elevation','Aspect','Slope','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology',
                'Horizontal_Distance_To_Roadways','Hillshade_9am','Hillshade_Noon','Hillshade_3pm','Horizontal_Distance_To_Fire_Points',
                'Wilderness_Area1','Wilderness_Area2','Wilderness_Area3','Wilderness_Area4','Soil_Type1','Soil_Type2',
                'Soil_Type3','Soil_Type4','Soil_Type5','Soil_Type6','Soil_Type7','Soil_Type8','Soil_Type9','Soil_Type10','Soil_Type11',
                'Soil_Type12','Soil_Type13','Soil_Type14','Soil_Type15','Soil_Type16','Soil_Type17','Soil_Type18','Soil_Type19',
                'Soil_Type20','Soil_Type21','Soil_Type22','Soil_Type23','Soil_Type24','Soil_Type25','Soil_Type26',
                'Soil_Type27','Soil_Type28','Soil_Type29','Soil_Type30','Soil_Type31','Soil_Type32','Soil_Type33',
                'Soil_Type34','Soil_Type35','Soil_Type36','Soil_Type37','Soil_Type38','Soil_Type39','Soil_Type40']]
y = forest['Cover_Type']

In [ ]:
# Divide X and y into training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

#### My computer keeps freezing at this cell: 

In [ ]:
# Train a linear SVM classifier
from sklearn.svm import SVC 
svm_model_linear = SVC(kernel = 'linear', C = 1).fit(X_train, y_train) 
svm_predictions = svm_model_linear.predict(X_test)

In [ ]:
# Create a model for accuracy for X_test   
accuracy_score = svm_model_linear.score(X_test, y_test)

## K Nearest Neighbor (KNN) Classifier

In [ ]:
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split

In [ ]:
# Read Data
forest = pd.read_csv('/Users/meganporter/OwnJupyterNotebooks/covtype.csv')

In [ ]:
# X is the classification predictor characteristics, y is the target label
X = forest[['Elevation','Aspect','Slope','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology',
                'Horizontal_Distance_To_Roadways','Hillshade_9am','Hillshade_Noon','Hillshade_3pm','Horizontal_Distance_To_Fire_Points',
                'Wilderness_Area1','Wilderness_Area2','Wilderness_Area3','Wilderness_Area4','Soil_Type1','Soil_Type2',
                'Soil_Type3','Soil_Type4','Soil_Type5','Soil_Type6','Soil_Type7','Soil_Type8','Soil_Type9','Soil_Type10','Soil_Type11',
                'Soil_Type12','Soil_Type13','Soil_Type14','Soil_Type15','Soil_Type16','Soil_Type17','Soil_Type18','Soil_Type19',
                'Soil_Type20','Soil_Type21','Soil_Type22','Soil_Type23','Soil_Type24','Soil_Type25','Soil_Type26',
                'Soil_Type27','Soil_Type28','Soil_Type29','Soil_Type30','Soil_Type31','Soil_Type32','Soil_Type33',
                'Soil_Type34','Soil_Type35','Soil_Type36','Soil_Type37','Soil_Type38','Soil_Type39','Soil_Type40']]
y = forest['Cover_Type']

In [ ]:
# Divide X, y into train and test data, input and output variables 
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

In [ ]:
# Train the KNN classifier 
from sklearn.neighbors import KNeighborsClassifier 
knn = KNeighborsClassifier(n_neighbors = 7).fit(X_train, y_train)

In [ ]:
# Accuracy on X_test 
accuracy_score = knn.score(X_test, y_test) 
print(accuracy_score) 

0.9652674987779942


#### The accuracy of KNN on this data set is 0.97

In [ ]:
# Create a confusion matrix/contingency table for visualization of accuracy 
knn_predictions = knn.predict(X_test)  
cm = confusion_matrix(y_test, knn_predictions)
print(cm)

[[51080  1717     1     0    35     5   105]
 [ 1537 69034    95     0   127    87    17]
 [    2   148  8566    28     9   137     0]
 [    0     1   144   520     0    39     0]
 [   22   244    15     0  2054     5     0]
 [    4   130   201    23     5  4021     0]
 [  139    22     0     0     1     0  4933]]


#### The values of the correct predictions for cover type span the diagonal entries from top-left to lower_right.

## Naive Bayes Classifier

In [ ]:
from sklearn.metrics import confusion_matrix 
from sklearn.model_selection import train_test_split

In [ ]:
# Read data
forest = pd.read_csv('/Users/meganporter/OwnJupyterNotebooks/covtype.csv')

In [ ]:
# X is the classification predictor characteristics, y is the target label
X = forest[['Elevation','Aspect','Slope','Horizontal_Distance_To_Hydrology','Vertical_Distance_To_Hydrology',
                'Horizontal_Distance_To_Roadways','Hillshade_9am','Hillshade_Noon','Hillshade_3pm','Horizontal_Distance_To_Fire_Points',
                'Wilderness_Area1','Wilderness_Area2','Wilderness_Area3','Wilderness_Area4','Soil_Type1','Soil_Type2',
                'Soil_Type3','Soil_Type4','Soil_Type5','Soil_Type6','Soil_Type7','Soil_Type8','Soil_Type9','Soil_Type10','Soil_Type11',
                'Soil_Type12','Soil_Type13','Soil_Type14','Soil_Type15','Soil_Type16','Soil_Type17','Soil_Type18','Soil_Type19',
                'Soil_Type20','Soil_Type21','Soil_Type22','Soil_Type23','Soil_Type24','Soil_Type25','Soil_Type26',
                'Soil_Type27','Soil_Type28','Soil_Type29','Soil_Type30','Soil_Type31','Soil_Type32','Soil_Type33',
                'Soil_Type34','Soil_Type35','Soil_Type36','Soil_Type37','Soil_Type38','Soil_Type39','Soil_Type40']]
y = forest['Cover_Type']

In [ ]:
# Divide X, y into train and test data, inputs and outputs 
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0)

In [ ]:
# Train the Naive Bayes classifier 
from sklearn.naive_bayes import GaussianNB 
gaussian_nbayes = GaussianNB().fit(X_train, y_train) 
gaussian_nbayes_predictions = gaussian_nbayes.predict(X_test)

In [ ]:
# Accuracy for the X_test 
accuracy_score = gaussian_nbayes.score(X_test, y_test) 
print(accuracy_score)

0.457883830282335


#### The accuracy for Naive Bayes on this data set is only 0.46.